include/pros/ext_adi.h

```diff
@@ -0,0 +1,613 @@
+ * failed, setting errno.
+ */
+int32_t ext_adi_encoder_shutdown(ext_adi_encoder_t enc);
+
+/**
+ * Reference type for an initialized ultrasonic.
+ *
+ * This merely contains the port number for the ultrasonic, unlike its use as an
+ * object to store encoder data in PROS 2.
+ */
+typedef int32_t ext_adi_ultrasonic_t;
+
+/**
+ * Gets the current ultrasonic sensor value in centimeters.
+ *
+ * If no object was found, zero is returned. If the ultrasonic sensor was never
+ * started, the return value is undefined. Round and fluffy objects can cause
+ * inaccurate values to be returned.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports, or the Smart Port is not within the valid range
+ * EADDRINUSE - The port is not configured as an ultrasonic
+ *
+ * \param ult
+ *        The adi_ultrasonic_t object from adi_ultrasonic_init() to read
+ *
+ * \return The distance to the nearest object in m^-4 (10000 indicates 1 meter),
+ * measured from the sensor's mounting points.
+ */
+int32_t ext_adi_ultrasonic_get(ext_adi_ultrasonic_t ult);
+
+/**
+ * Creates an ultrasonic object and configures the specified ports accordingly.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports, or the Smart Port is not within the valid range
+ * EADDRINUSE - The port is not configured as an ultrasonic
+ *
+ * \param smart_port
+ *        The smart port number that the ADI Expander is in
+ * \param adi_port_ping
+ *        The port connected to the orange OUTPUT cable. This should be in port
+ *        1, 3, 5, or 7 ('A', 'C', 'E', 'G').
+ * \param adi_port_echo
+ *        The port connected to the yellow INPUT cable. This should be in the
+ *        next highest port following port_ping.
+ *
+ * \return An adi_ultrasonic_t object to be stored and used for later calls to
+ * ultrasonic functions
+ */
+ext_adi_ultrasonic_t ext_adi_ultrasonic_init(uint8_t smart_port, uint8_t adi_port_ping, uint8_t adi_port_echo);
+
+/**
+ * Disables the ultrasonic sensor and voids the configuration on its ports.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports, or the Smart Port is not within the valid range
+ * EADDRINUSE - The port is not configured as an ultrasonic
+ *
+ * \param ult
+ *        The adi_ultrasonic_t object from adi_ultrasonic_init() to stop
+ *
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t ext_adi_ultrasonic_shutdown(ext_adi_ultrasonic_t ult);
+
+/**
+ * Reference type for an initialized gyroscope.
+ *
+ * This merely contains the port number for the gyroscope, unlike its use as an
+ * object to store encoder data in PROS 2.
```

Comment:
this should probably say "gyro" data
```suggestion
 * object to store gyro data in PROS 2.
```